okhttp/src/main/java/okhttp3/internal/PublicSuffixDatabase.java

```diff
@@ -0,0 +1,346 @@
+        exactMatch = match;
+        break;
+      }
+    }
+
+    // Wildcard rules throw a wrench into the mix. They are not restricted to appear in the leftmost
+    // position. An example is: *.*.foo. Therefore, we must exhaust the permutations,
+    // substituting a wildcard in the various label positions. For example, on the domain
+    // foo.bar.com, we'll try: [*, com], [foo, *, com], [*, bar, com], [*, *, com].
+    // Note: since we've already tried an exact match, we leave out those permutations.
+    String wildcardMatch = null;
+    if (domainLabelsUtf8Bytes.length > 1) {
+      // It seems highly unlikely (impossible?) the rightmost label will ever be a wildcard.
+      byte[][] labelsWithWildcard = new byte[domainLabelsUtf8Bytes.length][];
+      labelsWithWildcard[domainLabelsUtf8Bytes.length - 1] =
+          domainLabelsUtf8Bytes[domainLabelsUtf8Bytes.length - 1];
+
+      int firstLevel = domainLabelsUtf8Bytes.length - 1;
+      outer: for (int currentLevel = 0; currentLevel < firstLevel; currentLevel++) {
+        int numberOfWildcards = firstLevel - currentLevel;
+        int permutations = 1 << numberOfWildcards;
+        for (int i = 1; i < permutations; i++) {
+          for (int labelPos = firstLevel - 1; labelPos >= currentLevel; labelPos--) {
+            // Use the bit positions to indicate whether we should place a wildcard there. For
+            // example with foo.bar.com: 01, 10, 11. The second level is position 1, third is
+            // position 2, etc.
+            int labelBitPos = 1 << firstLevel - 1 - labelPos;
+            int useWildcard = labelBitPos & i;
+
+            labelsWithWildcard[labelPos] = useWildcard > 0
+                ? WILDCARD_LABEL
+                : domainLabelsUtf8Bytes[labelPos];
+          }
+          String match = binarySearchBytes(publicSuffixListBytes, labelsWithWildcard, currentLevel);
+          if (match != null) {
+            wildcardMatch = match;
+            break outer;
+          }
+        }
+      }
+    }
+
+    // Exception rules only apply to wildcard rules, so only try it if we matched a wildcard.
+    String exception = null;
+    if (wildcardMatch != null) {
+      for (int i = 0; i < domainLabelsUtf8Bytes.length - 1; i++) {
+        String rule = binarySearchBytes(publicSuffixExceptionListBytes, domainLabelsUtf8Bytes, i);
+        if (rule != null) {
+          exception = rule;
+          break;
+        }
+      }
+    }
+
+    if (exception != null) {
+      // Signal we've identified an exception rule.
+      exception = "!" + exception;
+      return exception.split("\\.");
+    }
+
+    if (exactMatch == null && wildcardMatch == null) return null;
+
+    String[] exactRuleLabels = exactMatch != null
+        ? exactMatch.split("\\.")
+        : EMPTY_RULE;
+
+    String[] wildcardRuleLabels = wildcardMatch != null
+        ? wildcardMatch.split("\\.")
+        : EMPTY_RULE;
+
+    return exactRuleLabels.length > wildcardRuleLabels.length
```

Comment:
Ohh, interesting point. The specification mentions that a wildcard is for an entire level, i.e. it shouldn't be possible for `cash.square.com` and `*.square.com` to be in the list. Maybe there is a better way of expressing that?